reactos/drivers/usb/usbport/queue.c

```diff
@@ -0,0 +1,1389 @@
+
+    while (TRUE)
+    {
+        if (IsListEmpty(&List))
+            break;
+
+        Endpoint = CONTAINING_RECORD(List.Flink,
+                                     USBPORT_ENDPOINT,
+                                     FlushLink);
+
+        RemoveHeadList(&List);
+
+        Endpoint->FlushLink.Flink = NULL;
+        Endpoint->FlushLink.Blink = NULL;
+
+        if (!IsListEmpty(&Endpoint->PendingTransferList))
+        {
+            USBPORT_FlushPendingTransfers(Endpoint);
+        }
+    }
+
+    DPRINT_CORE("USBPORT_FlushAllEndpoints: exit\n");
+}
+
+ULONG
+NTAPI
+USBPORT_KillEndpointActiveTransfers(IN PDEVICE_OBJECT FdoDevice,
+                                    IN PUSBPORT_ENDPOINT Endpoint)
+{
+    PLIST_ENTRY ActiveList;
+    PUSBPORT_TRANSFER Transfer;
+    ULONG KilledTransfers = 0;
+
+    DPRINT_CORE("USBPORT_KillEndpointActiveTransfers \n");
+
+    ActiveList = Endpoint->TransferList.Flink;
+
+    if (!IsListEmpty(&Endpoint->TransferList))
+    {
+        while (ActiveList && ActiveList != &Endpoint->TransferList)
+        {
+            ++KilledTransfers;
+
+            Transfer = CONTAINING_RECORD(ActiveList,
+                                         USBPORT_TRANSFER,
+                                         TransferLink);
+
+            Transfer->Flags |= TRANSFER_FLAG_ABORTED;
+
+            ActiveList = Transfer->TransferLink.Flink;
+        }
+    }
+
+    USBPORT_FlushPendingTransfers(Endpoint);
+    USBPORT_FlushCancelList(Endpoint);
+
+    return KilledTransfers;
+}
+
+VOID
+NTAPI
+USBPORT_FlushController(IN PDEVICE_OBJECT FdoDevice)
+{
+    PUSBPORT_DEVICE_EXTENSION  FdoExtension;
+    PLIST_ENTRY Entry;
+    PUSBPORT_ENDPOINT Endpoint;
+    ULONG KilledTransfers = 0;
+    PLIST_ENTRY EndpointList;
+    KIRQL OldIrql;
+    LIST_ENTRY FlushList;
+
+    DPRINT_CORE("USBPORT_FlushController \n");
+
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)FdoDevice->DeviceExtension;
+
+    EndpointList = &FdoExtension->EndpointList;
+
+    while (TRUE)
+    {
+        KeAcquireSpinLock(&FdoExtension->EndpointListSpinLock, &OldIrql);
+
+        InitializeListHead(&FlushList);
+
+        Entry = EndpointList->Flink;
+
+        if (!IsListEmpty(EndpointList))
+        {
+            while (Entry && Entry != EndpointList)
+            {
+                Endpoint = CONTAINING_RECORD(Entry,
+                                             USBPORT_ENDPOINT,
+                                             EndpointLink);
+
+                if (Endpoint->StateLast != USBPORT_ENDPOINT_CLOSED &&
+                    Endpoint->StateLast != USBPORT_ENDPOINT_NOT_HANDLED)
+                {
+                    InterlockedIncrement(&Endpoint->LockCounter);
+                    InsertTailList(&FlushList, &Endpoint->FlushControllerLink);
+                }
+
+                Entry = Endpoint->EndpointLink.Flink;
+            }
+        }
+
+        KeReleaseSpinLock(&FdoExtension->EndpointListSpinLock, OldIrql);
+
+        if (IsListEmpty(&FlushList))
```

Comment:
?
VOID
NTAPI
USBPORT_FlushController(IN PDEVICE_OBJECT FdoDevice)
{
    PUSBPORT_DEVICE_EXTENSION  FdoExtension;
    PLIST_ENTRY Entry;
    PUSBPORT_ENDPOINT Endpoint;
    ULONG KilledTransfers = 0;